### ISSUE TITLE
Strange error

### ISSUE THREAD

**Issue by Gredys** [sidorares/node-mysql2#178]
I am getting strange error with handling this query
```
SELECT `ac`.`id`, `ac`.`username`, `ac`.`auth_code`, `ac`.`access_token`, `ac`.`refresh_token`, `acs`.`session`, CONCAT('[', GROUP_CONCAT(DISTINCT `acf`.`flag` SEPARATOR ','), ']') flags FROM account ac LEFT JOIN account_flags acf ON `acf`.account = `ac`.id LEFT JOIN account_session acs ON `acs`.account = `ac`.id WHERE `acs`.`session`=?
```
Just for testing i passed to '?' this string 'asid=75efb145482ce22f4544390cad233c749c1b43e4' and got this error
```
1434548751936 - Origin: Worker (PID 4648)
    [Error] RangeError: out of range index
    at RangeError (native)
    at Packet.readLengthCodedString (..\node_modules\mysql2\lib\packets\packet.js:271:22)
    at new BinaryRow (evalmachine.<anonymous>:8:29)
    at Execute.Query.row (..\node_modules\mysql2\lib\commands\query.js:195:13)
    at Execute.Command.execute (..\node_modules\mysql2\lib\commands\command.js:34:20)
    at Connection.handlePacket (..\node_modules\mysql2\lib\connection.js:310:28)
    at PacketParser.onPacket (..\node_modules\mysql2\lib\connection.js:83:65)
    at PacketParser.executeStart (..\node_modules\mysql2\lib\packet_parser.js:39:12)
    at Socket.<anonymous> (..\node_modules\mysql2\lib\connection.js:95:31)
    at Socket.emit (events.js:107:17)
```
Here core, that runs this query (from wrapper)
```
Database = {
...
    Query: function(db, query, args, callback) {
        this.Connect(db).execute(query, args || [], function(err, rows) {
            callback(err, rows.length != 0 ? rows : null);
        });
    },
...
}
```
I am also using SocketCluster and Wait.For. Any idea how to fix this problem?

P.S. All working great  if pass '75efb145482ce22f4544390cad233c749c1b43e4' to '?'.
P.S.S. I figured out, that offset from method Packet:readLengthCodedNumberExt will return always 'BN: 22000' (from tag 0xfe)

**Comment by sidorares** [comment]
Could you show log when you connect with `createConnection({ debug: true });` ?

**Comment by Gredys** [comment]
Sure, here. I tried to disable all other queries, so there must be only data from broken one.
```
On connection

Add command: Connection
0 undefined ==> Connection#handshakeInit(0,,74)
 raw: 0a352e362e3234005f000000253e49356c413e5400fff70802007f8015000000000000000000006d485d7a6e396e403425586f006d7973716c5f6e61746976655f70617373776f726400
Server hello packet: capability flags:-2139097089=(long password, found rows, long flag, connect with db, no schema, compress, odbc, local files, ignore space, protocol 41, interactive, ignore sigpipe, transactions, reserved, secure connection, multi statements, multi results, ps multi results, plugin auth, remember options)
Sending handshake packet: flags:193487=(long password, found rows, long flag, connect with db, odbc, local files, ignore space, protocol 41, ignore sigpipe, transactions, reserved, secure connection, multi results)
0 95 <== Connection#handshakeInit(1,,70)
0 95 ==> Connection#handshakeResult(2,maybeOK,7)
 raw: 00000002000000
```
```
On query

Add command: prepare
0 95 <== prepare#start(0,,343)
0 95 ==> prepare#prepareHeader(1,maybeOK,12)
 raw: 000100000007000100000000
0 95 ==> prepare#readParameter(2,,23)
 raw: 03646566000000013f000c3f0000000000fd8000000000
0 95 ==> prepare#parametersEOF(3,EOF,5)
 raw: fe00000200
0 95 ==> prepare#readField(4,,42)
 raw: 036465660764657673697465026163076163636f756e740269640269640c3f000b000000030342000000
0 95 ==> prepare#readField(5,,54)
 raw: 036465660764657673697465026163076163636f756e7408757365726e616d6508757365726e616d650c2100c0000000fd0110000000
0 95 ==> prepare#readField(6,,56)
 raw: 036465660764657673697465026163076163636f756e7409617574685f636f646509617574685f636f64650c21005a000000fd0110000000
0 95 ==> prepare#readField(7,,62)
 raw: 036465660764657673697465026163076163636f756e740c6163636573735f746f6b656e0c6163636573735f746f6b656e0c21005a000000fd0110000000
0 95 ==> prepare#readField(8,,64)
 raw: 036465660764657673697465026163076163636f756e740d726566726573685f746f6b656e0d726566726573685f746f6b656e0c2100fd020000fc1110000000
0 95 ==> prepare#readField(9,,61)
 raw: 036465660764657673697465036163730f6163636f756e745f73657373696f6e0773657373696f6e0773657373696f6e0c210056010000fd0010000000
0 95 ==> prepare#readField(10,,27)
 raw: 0364656600000005666c616773000c210005040000fd00001f0000
0 95 ==> prepare#fieldsEOF(11,EOF,5)
 raw: fe00000200
Add command: executeStatement
0 95 <== executeStatement#start(0,,64)
0 95 ==> executeStatement#resultsetHeader(1,,1)
 raw: 07
        Resultset header received, expecting 7 column definition packets
0 95 ==> executeStatement#readField(2,,42)
 raw: 036465660764657673697465026163076163636f756e740269640269640c3f000b000000030342000000
0 95 ==> executeStatement#readField(3,,54)
 raw: 036465660764657673697465026163076163636f756e7408757365726e616d6508757365726e616d650c2100c0000000fd0110000000
0 95 ==> executeStatement#readField(4,,56)
 raw: 036465660764657673697465026163076163636f756e7409617574685f636f646509617574685f636f64650c21005a000000fd0110000000
0 95 ==> executeStatement#readField(5,,62)
 raw: 036465660764657673697465026163076163636f756e740c6163636573735f746f6b656e0c6163636573735f746f6b656e0c21005a000000fd0110000000
0 95 ==> executeStatement#readField(6,,64)
 raw: 036465660764657673697465026163076163636f756e740d726566726573685f746f6b656e0d726566726573685f746f6b656e0c2100fd020000fc1110000000
0 95 ==> executeStatement#readField(7,,61)
 raw: 036465660764657673697465036163730f6163636f756e745f73657373696f6e0773657373696f6e0773657373696f6e0c210056010000fd0010000000
0 95 ==> executeStatement#readField(8,,27)
 raw: 0364656600000005666c616773000c210005040000fd00001f0000
0 95 ==> executeStatement#fieldsEOF(9,EOF,5)
 raw: fe00002200
Compiled binary protocol row parser:
(function(){ return function BinaryRow(packet) {
  var statusByte = packet.readInt8();
  var nullBitmaskByte0 = packet.readInt8();
  var nullBitmaskByte1 = packet.readInt8();
  // "id": LONG
  this["id"] = packet.readSInt32();
  // "username": VAR_STRING
  this["username"] = packet.readLengthCodedString();
  // "auth_code": VAR_STRING
  this["auth_code"] = packet.readLengthCodedString();
  // "access_token": VAR_STRING
  this["access_token"] = packet.readLengthCodedString();
  // "refresh_token": BLOB
  this["refresh_token"] = packet.readLengthCodedString();
  // "session": VAR_STRING
  if (nullBitmaskByte0 & 128)
    this["session"] = null;
  else
    this["session"] = packet.readLengthCodedString();
  // "flags": VAR_STRING
  if (nullBitmaskByte1 & 1)
    this["flags"] = null;
  else
    this["flags"] = packet.readLengthCodedString();
}; })()
0 95 ==> executeStatement#row(10,maybeOK,3)
 raw: 00fc01
1434550563821 - Origin: Worker (PID 5812)
    [Error] RangeError: out of range index
    at RangeError (native)
    at Packet.readLengthCodedString (..\node_modules\mysql2\lib\packets\packet.js:271:22)
    at new BinaryRow (evalmachine.<anonymous>:8:29)
    at Execute.Query.row (..\node_modules\mysql2\lib\commands\query.js:194:13)
    at Execute.Command.execute (..\node_modules\mysql2\lib\commands\command.js:34:20)
    at Connection.handlePacket (..\node_modules\mysql2\lib\connection.js:310:28)
    at PacketParser.onPacket (..\node_modules\mysql2\lib\connection.js:83:65)
    at PacketParser.executeStart (..\node_modules\mysql2\lib\packet_parser.js:39:12)
    at Socket.<anonymous> (..\node_modules\mysql2\lib\connection.js:95:31)
    at Socket.emit (events.js:107:17)
1434550563834 - Worker 0 exited - Exit code: 0
```

**Comment by sidorares** [comment]
Thanks! This might be enough for me to debug

**Comment by sidorares** [comment]
actually this is not enough :( I can see there is short 3-bytes packet that does not match parser but not sure why. Is it possible to reduce test case to something I can replicate?

**Comment by Gredys** [comment]
Okay. I will try in few minutes and tell u results.

**Comment by Gredys** [comment]
Here link to sql file that contains all tables - https://dl.dropboxusercontent.com/u/20012031/account.sql

Code that will cause this error. Tried to simplify sql, but error dissapeared.
```
var mysql = require('mysql2');

var connection = mysql.createConnection({
    debug    : true,
    host     : 'localhost',
    user     : 'root',
    password : '',
    database : 'devsite'
});

connection.connect(function(err) {
    if (err) {
        console.error(err);
        return;
    }

    connection.execute("SELECT `ac`.`id`, `ac`.`username`, `ac`.`auth_code`, `ac`.`access_token`, `ac`.`refresh_token`, `acs`.`session`, CONCAT('[', GROUP_CONCAT(DISTINCT `acf`.`flag` SEPARATOR ','), ']') flags FROM account ac LEFT JOIN account_flags acf ON `acf`.account = `ac`.id LEFT JOIN account_session acs ON `acs`.account = `ac`.id WHERE `acs`.`session`=?", ["asid=75efb145482ce22f4544390cad233c749c1b43e4"], function(err, rows) {
        console.log(err, rows);
    });
});
```

And again error, but now its a little bit diffient (cos I am not using SocketCluster and Wait.For now)
```
Add command: Connection
0 undefined ==> Connection#handshakeInit(0,,74)
 raw: 0a352e362e3234006b0000007434433726596e3e00fff70802007f8015000000000000000000002c5d703e2230694350485967006d7973716c5f6e61746976655f7061
7373776f726400
Server hello packet: capability flags:-2139097089=(long password, found rows, long flag, connect with db, no schema, compress, odbc, local f
iles, ignore space, protocol 41, interactive, ignore sigpipe, transactions, reserved, secure connection, multi statements, multi results, ps
 multi results, plugin auth, remember options)
Sending handshake packet: flags:193487=(long password, found rows, long flag, connect with db, odbc, local files, ignore space, protocol 41,
 ignore sigpipe, transactions, reserved, secure connection, multi results)
0 107 <== Connection#handshakeInit(1,,70)
0 107 ==> Connection#handshakeResult(2,maybeOK,7)
 raw: 00000002000000
Add command: prepare
0 107 <== prepare#start(0,,343)
0 107 ==> prepare#prepareHeader(1,maybeOK,12)
 raw: 000100000007000100000000
0 107 ==> prepare#readParameter(2,,23)
 raw: 03646566000000013f000c3f0000000000fd8000000000
0 107 ==> prepare#parametersEOF(3,EOF,5)
 raw: fe00000200
0 107 ==> prepare#readField(4,,42)
 raw: 036465660764657673697465026163076163636f756e740269640269640c3f000b000000030342000000
0 107 ==> prepare#readField(5,,54)
 raw: 036465660764657673697465026163076163636f756e7408757365726e616d6508757365726e616d650c2100c0000000fd0110000000
0 107 ==> prepare#readField(6,,56)
 raw: 036465660764657673697465026163076163636f756e7409617574685f636f646509617574685f636f64650c21005a000000fd0110000000
0 107 ==> prepare#readField(7,,62)
 raw: 036465660764657673697465026163076163636f756e740c6163636573735f746f6b656e0c6163636573735f746f6b656e0c21005a000000fd0110000000
0 107 ==> prepare#readField(8,,64)
 raw: 036465660764657673697465026163076163636f756e740d726566726573685f746f6b656e0d726566726573685f746f6b656e0c2100fd020000fc1110000000
0 107 ==> prepare#readField(9,,61)
 raw: 036465660764657673697465036163730f6163636f756e745f73657373696f6e0773657373696f6e0773657373696f6e0c210056010000fd0010000000
0 107 ==> prepare#readField(10,,27)
 raw: 0364656600000005666c616773000c210005040000fd00001f0000
0 107 ==> prepare#fieldsEOF(11,EOF,5)
 raw: fe00000200
Add command: executeStatement
0 107 <== executeStatement#start(0,,64)
0 107 ==> executeStatement#resultsetHeader(1,,1)
 raw: 07
        Resultset header received, expecting 7 column definition packets
0 107 ==> executeStatement#readField(2,,42)
 raw: 036465660764657673697465026163076163636f756e740269640269640c3f000b000000030342000000
0 107 ==> executeStatement#readField(3,,54)
 raw: 036465660764657673697465026163076163636f756e7408757365726e616d6508757365726e616d650c2100c0000000fd0110000000
0 107 ==> executeStatement#readField(4,,56)
 raw: 036465660764657673697465026163076163636f756e7409617574685f636f646509617574685f636f64650c21005a000000fd0110000000
0 107 ==> executeStatement#readField(5,,62)
 raw: 036465660764657673697465026163076163636f756e740c6163636573735f746f6b656e0c6163636573735f746f6b656e0c21005a000000fd0110000000
0 107 ==> executeStatement#readField(6,,64)
 raw: 036465660764657673697465026163076163636f756e740d726566726573685f746f6b656e0d726566726573685f746f6b656e0c2100fd020000fc1110000000
0 107 ==> executeStatement#readField(7,,61)
 raw: 036465660764657673697465036163730f6163636f756e745f73657373696f6e0773657373696f6e0773657373696f6e0c210056010000fd0010000000
0 107 ==> executeStatement#readField(8,,27)
 raw: 0364656600000005666c616773000c210005040000fd00001f0000
0 107 ==> executeStatement#fieldsEOF(9,EOF,5)
 raw: fe00002200
Compiled binary protocol row parser:
(function(){ return function BinaryRow(packet) {
  var statusByte = packet.readInt8();
  var nullBitmaskByte0 = packet.readInt8();
  var nullBitmaskByte1 = packet.readInt8();
  // "id": LONG
  this["id"] = packet.readSInt32();
  // "username": VAR_STRING
  this["username"] = packet.readLengthCodedString();
  // "auth_code": VAR_STRING
  this["auth_code"] = packet.readLengthCodedString();
  // "access_token": VAR_STRING
  this["access_token"] = packet.readLengthCodedString();
  // "refresh_token": BLOB
  this["refresh_token"] = packet.readLengthCodedString();
  // "session": VAR_STRING
  if (nullBitmaskByte0 & 128)
    this["session"] = null;
  else
    this["session"] = packet.readLengthCodedString();
  // "flags": VAR_STRING
  if (nullBitmaskByte1 & 1)
    this["flags"] = null;
  else
    this["flags"] = packet.readLengthCodedString();
}; })()
0 107 ==> executeStatement#row(10,maybeOK,3)
 raw: 00fc01
..\node_modules\mysql2\lib\packets\packet.js:271
  return this.buffer.utf8Slice(this.offset - len, this.offset);
                     ^
RangeError: out of range index
    at RangeError (native)
    at Packet.readLengthCodedString (..\node_modules\mysql2\lib\packets\packet.js:271:22)
    at new BinaryRow (evalmachine.<anonymous>:8:29)
    at Execute.Query.row (..\node_modules\mysql2\lib\commands\query.js:194:13)
    at Execute.Command.execute (..\node_modules\mysql2\lib\commands\command.js:34:20)
    at Connection.handlePacket (..\node_modules\mysql2\lib\connection.js:310:28)
    at PacketParser.onPacket (..\node_modules\mysql2\lib\connection.js:83:65)
    at PacketParser.executeStart (..\node_modules\mysql2\lib\packet_parser.js:39:12)
    at Socket.<anonymous> (..\node_modules\mysql2\lib\connection.js:95:31)
    at Socket.emit (events.js:107:17)
```

**Comment by sidorares** [comment]
great, I can replicate it. Thanks

**Comment by sidorares** [comment]
the problem is that server sends 
```
{ id: null,
    username: null,
    auth_code: null,
    access_token: null,
    refresh_token: null,
    session: null,
    flags: null }
```
row, while in the schema fields are marked as "Not null". I did not expect such behaviour from mysql and [this test](https://github.com/sidorares/node-mysql2/blob/master/lib/compile_binary_parser.js#L60-L62) does not work. I guess I need to remove this optimisation even if it work in most cases

**Comment by sidorares** [comment]
so "00fc01" row is "status == 00", followed by 7 nulls ( encoded by two bytes 0xfc 0x01, or 11111100 +  0000001 - in the first byte first two bits reserved )

**Comment by Gredys** [comment]
Its returning null because i am using `group concate` with joins.
Even if rows will be seted to `not null` it will return null if there no any data.

**Comment by sidorares** [comment]
Still its very strange to see: column type: not null. Value: null

**Comment by sidorares** [comment]
published as v0.15.7
